source/server/server.cc

```diff
@@ -237,16 +239,25 @@ MetricSnapshotImpl::MetricSnapshotImpl(Stats::Store& store, TimeSource& time_sou
         text_readouts_.push_back(text_readout);
       });
 
+  Upstream::HostUtility::forEachHostMetric(
```

Comment:
we want to be performant with 10k clusters. Assuming we never use this feature, I think this model still requires us to visit all 10k clusters on every stats sink.

WDYT of keeping the clusters with per-endpoint stats enabled in a std::set in the cluster manager, and iterate over that set? That would enable good performance with 10k clustesr of which only 1 has per-endpoint stats.

A compromise is to just keep a count of clusters with this enabled, and early-exit the loop inside forEachHostMetric using that. That would probably be fine in practice if users with 10k clusters never used this feature.

Reply:
I thought about this, and my conclusion was that the amount of work to publish the other stats for 10k clusters is >>>>> greater than checking one enabled() function on each cluster. Eg, even if there was a ton of filtering so that each cluster had ~5 stats (but 10k of them), the string manipulations needed to publish the 50000 metrics should make this check unnoticable.

And thats' what `stats_handler_speed_test` tests, and it's not showing any meaingful slowdown. 

I'm worried about the complexity of trying to keep track of how many clusters have this enabled; we'd need to catch every path where cluster config can be modified (in addition to all add/remove operations). Is it worth the complexity and potential for bugs?

Reply:
Good point! resolving.